resample/tests/test_bootstrap.py

```diff
@@ -222,3 +225,51 @@ def test_confidence_interval_invalid_ci_method_raises():
     msg = "method must be 'percentile', 'student', or 'bca'"
     with pytest.raises(ValueError, match=msg):
         confidence_interval(np.mean, (1, 2, 3), ci_method="foobar")
+
+
+@pytest.mark.parametrize("method", ("ordinary", "balanced"))
+def test_bias_on_unbiased(method, rng):
+    data = (0, 1, 2, 3)
+    r = bias(np.mean, data, method="balanced", random_state=rng)
+
+    if method == "balanced":
+        # bias is exactly zero for linear functions with the balanced bootstrap
+        assert r == 0
+    else:
+        # bias is not exactly zero for ordinary bootstrap
+        assert r == pytest.approx(0)
+
+
+@pytest.mark.parametrize("method", ("ordinary", "balanced"))
+def test_bias_on_biased(method, rng):
+    def biased(x):
+        return np.var(x, ddof=0)
+
+    data = np.arange(100)
+    bad = biased(data)
+    correct = np.var(data, ddof=1)
+
+    r = bias(biased, data, method=method, size=10000, random_state=rng)
+    true_bias = bad - correct
+    assert r == pytest.approx(true_bias, rel=0.1)
```

Comment:
Hmm, not sure if we should call this true_bias since we're really taking the difference between biased and unbiased _estimates_ which isn't the same thing as bias.

We could alternatively do something like setting data = np.random.randn(N) and test whether our estimate of bias for np.var(data, ddof=0) is close to 1 / N (which _is_ the true bias).

Reply:
The bootstrap considers the empirical CDF as the true CDF and draws samples from that. The variance of the empirical  CDF is exactly `np.var(data, ddof=1)`. I think `true_bias` is correct, no?

Reply:
I think the original data sample should be fixed when we investigate properties of the resampling algorithms.

Reply:
The variance of the empirical distribution is exactly np.var(data, ddof=0) not np.var(data, ddof=1), so under this interpretation what you are calling bad is actually the true value. But this seems to me like the wrong way to look at it. When we're using the bootstrap we don't pretend to have a complete census of the underlying population in our sample; if we did we wouldn't need to talk about bias. It only makes sense in the context of a larger distribution, which was why I was recommending we use one where we know the true properties of our estimators, including their bias.

Reply:
> The variance of the empirical distribution is exactly np.var(data, ddof=0) not np.var(data, ddof=1), so under this interpretation what you are calling bad is actually the true value.

Ok, I see the point.

> It only makes sense in the context of a larger distribution, which was why I was recommending we use one where we know the true properties of our estimators, including their bias.

We don't talk about the bias of the data distribution, this does not have bias. We talk about a bias of an estimator applied to samples from that distribution. This can be biased even if the distribution is not. The empirical CDF on which the bootstrap is based is an unbiased maximum-likelihood estimate of the true CDF.

I wrote these statistical tests of the bootstrap, and there I found it to be important to distinguish between the original distribution and the distribution from which we sample. Example: if the true distribution is standard normal, then we know its mean and variance. But for a particular finite sample, the mean and variance of that sample will be different. When computing the bias of e.g. the arithmetic mean, we need to use the actual sample mean of the sample, not the original theoretical mean of zero.

Reply:
> Ok, I see the point.

I just tried `np.var(data, ddof=0)` and then the test fails by a large margin. Something must be wrong with your argument.

Reply:
My argument wasn't to replace np.var(data, ddof=0) in the existing test; the argument was that as written this test seems to confuse the definition of bias itself. In it we are making an assertion about our bias estimate being close to the true bias, _but we have no way of knowing what this true bias is_. We have a sample but no knowledge of the distribution from which it arose, and hence no way to make assertions about what our bias estimate should be.

Therefore I was recommending that we instead use a scenario where we _know what is it we are estimating_, such as when estimating the bias of the variance estimator np.var(data, ddof=0) when sampling from a standard normal:
```python
data = np.random.randn(100)
result = bias(fn=lambda x: np.var(x, ddof=0), sample=data, size=10_000)
assert result == pytest.approx(-0.01, rel=0.01)
```
Note we're deriving the bias _a priori_, not from the available data since this is not possible (if it were then there would be no reason to estimate it).

Reply:
I understand most of this, but the bias of `np.var(x, ddof=0)` is known for any distribution, it is independent of the distribution. The bias is also proportional to the variance of `data`, which is what I use in my test. I use the same argument again my presentation for PyHEP.

Reply:
> I understand most of this, but the bias of `np.var(x, ddof=0)` is known for any distribution, it is independent of the distribution. The bias is also proportional to the variance of `data`

This is not correct, the bias is - sigma^2 / n, where sigma^2 is the (unknown) variance of the underlying distribution. It's not a function of the data (except through the sample size).

Reply:
Neither the bootstrap nor the jackknife can compute the expected bias. The best they can do is to correctly compute the bias for the particular data sample at hand. This is what I do and that is why I think the test is correct.

What you are suggesting is to compare to the expected bias. This may work if you chose a lucky seed and/or make the tolerance large enough. But it is a less sensitive test, because - again - neither the bootstrap nor the jackknife are able to estimate the true bias. Try this code for yourself

```
from resample.bootstrap import bias
import numpy as np
import matplotlib.pyplot as plt

biased = lambda x: np.var(x, ddof=0)
unbiased = lambda x: np.var(x, ddof=1)

delta1 = []
delta2 = []
for i in range(100):
    data = np.random.randn(100)
    result = bias(biased, data, size=1000)
    delta1.append(result - 1/len(data))
    delta2.append(result - (biased(data) - unbiased(data)))

print(np.std(delta1), np.std(delta2))
```
On my compute, it prints: 0.0014394905080510607 0.0004251280541256813

Reply:
AFAIU, `np.std(delta2)` will converge to zero if the number of bootstrap replicas is increased, while `np.std(delta1)` will converge to a non-zero value.

Reply:
Can we please close this topic now?

Reply:
If we aren't attempting to estimate the true bias within some tolerance then we shouldn't call it true_bias in the test (which is evidently what is causing the confusion)